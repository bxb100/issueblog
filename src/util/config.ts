import * as core from '@actions/core'
import * as z from 'zod'

// schema
const keys = [
    'github_token',
    'md_header',
    'issue_number',
    'recent_limit',
    'anchor_number',
    'links_title',
    'recent_title',
    'top_title',
    'unlabeled_title',
    'blog_image_url'
]

const commonConfigSchema = z.object({
    github_token: z.string(),
    md_header: z.string(),
    issue_number: z.string().optional(),
    recent_limit: z.string().default('5'),
    anchor_number: z.string().default('5'),
    links_title: z.string(),
    recent_title: z.string(),
    top_title: z.string(),
    unlabeled_title: z.string(),
    blog_image_url: z.string().default('blog.png')
})

export type Config = z.infer<typeof commonConfigSchema>

/**
 * 将 action.yml 中的 input 入参转换为对象
 *
 * @returns {Config} 映射的对象
 */
export function getConfig(): Config {
    const raw: {[key: string]: string} = {}

    for (const key of keys) {
        const v = core.getInput(key)
        if (v) {
            raw[key] = v
        }
    }

    core.debug(`Raw config: ${JSON.stringify(raw)}`)
    return commonConfigSchema.parse(raw)
}
