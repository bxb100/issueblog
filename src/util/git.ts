import * as core from '@actions/core'
import {GitStatus} from '../common/types/git-status'
import {exec} from '@actions/exec'
import path from 'path'
import {statSync} from 'fs'

export async function gitStatus(): Promise<GitStatus[]> {
    core.debug('Getting gitStatus()')
    let output = ''
    await exec('git', ['status', '-s'], {
        listeners: {
            stdout: (data: Buffer) => {
                output += data.toString()
            }
        }
    })
    core.debug(`=== output was:\n${output}`)
    return output
        .split('\n')
        .map(l => l.trim())
        .filter(l => l !== '')
        .map(l => {
            const chunks = l.split(/\s+/)
            return {
                flag: chunks[0],
                path: chunks[1]
            } as GitStatus
        })
}

async function getHeadSize(filePath: string): Promise<number | undefined> {
    let raw = ''
    const exitCode = await exec('git', ['cat-file', '-s', `HEAD:${filePath}`], {
        listeners: {
            stdline: (data: string) => {
                raw += data
            }
        }
    })
    core.debug(`raw cat-file output: ${exitCode} '${raw}'`)
    if (exitCode === 0) {
        return parseInt(raw, 10)
    }
}

async function diffSize(file: GitStatus): Promise<number> {
    switch (file.flag) {
        case 'M': {
            const stat = statSync(file.path)
            core.debug(
                `Calculating diff for ${JSON.stringify(file)}, with size ${
                    stat.size
                }b`
            )

            // get old size and compare
            const oldSize = await getHeadSize(file.path)
            const delta =
                oldSize === undefined ? stat.size : stat.size - oldSize
            core.debug(
                ` ==> ${file.path} modified: old ${oldSize}, new ${stat.size}, delta ${delta}b `
            )
            return delta
        }
        case 'A': {
            const stat = statSync(file.path)
            core.debug(
                `Calculating diff for ${JSON.stringify(file)}, with size ${
                    stat.size
                }b`
            )

            core.debug(` ==> ${file.path} added: delta ${stat.size}b`)
            return stat.size
        }
        case 'D': {
            const oldSize = await getHeadSize(file.path)
            const delta = oldSize === undefined ? 0 : oldSize
            core.debug(` ==> ${file.path} deleted: delta ${delta}b`)
            return delta
        }
        default: {
            throw new Error(
                `Encountered an unexpected file status in git: ${file.flag} ${file.path}`
            )
        }
    }
}

export async function diff(filename: string): Promise<number> {
    const statuses = await gitStatus()
    core.debug(
        `Parsed statuses: ${statuses.map(s => JSON.stringify(s)).join(', ')}`
    )
    const status = statuses.find(s => path.relative(s.path, filename) === '')
    if (typeof status === 'undefined') {
        core.info(`No status found for ${filename}, aborting.`)
        return 0 // there's no change to the specified file
    }
    return await diffSize(status)
}

const lsFile = async (
    commandLine: string,
    args?: string[]
): Promise<string[]> => {
    const raw: string[] = []
    await exec(commandLine, args, {
        listeners: {
            stdline: (data: string) => {
                raw.push(data.trim())
            }
        }
    })
    return raw.filter(l => l !== '')
}

export async function getUnstagedFiles(): Promise<string[]> {
    return lsFile('git', ['ls-files', '--others', '--exclude-standard'])
}

export async function getModifiedUnstagedFiles(): Promise<string[]> {
    return lsFile('git', ['ls-files', '-m'])
}

export async function submodulePath(): Promise<string[]> {
    const raw: string[] = []
    await exec('git', ['ls-files', '-s'], {
        listeners: {
            stdline: (data: string) => {
                if (data.trim().startsWith('160000')) {
                    raw.push(data.trim().split('\t')[1])
                }
            }
        }
    })
    return raw
}
