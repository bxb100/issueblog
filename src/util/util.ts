import {IComment} from '../common/interface/comment'
import {IIssue} from '../common/interface/issue'
import {IsoDateString} from '../common/types/iso-date-string'
import {Issue} from '../common/clazz/issue'
import {MetadataInfo} from '../common/types/metadata'
import {WrapperToString} from '../common/types/wrapper-to-string'

export function isOwnBy(obj: IIssue | IComment, username: string): boolean {
    return obj.user?.login === username
}

export function wrapDetails<T>(
    shows: T[],
    hides: T[],
    wrapper: WrapperToString<T>
): string {
    let result = ''

    if (shows.length !== 0) {
        result += shows.map(wrapper).join('')
    }

    if (hides.length !== 0) {
        result += '<details><summary>MORE</summary>'
        result += '\n\n'

        result += hides.map(wrapper).join('')

        result += '</details>'
        result += '\n'
    }
    result += '\n'
    return result
}

export function backupFileName(issue: Issue): string {
    return `${issue.number}-${issue.title.replace(/\t|\s/g, '_')}.md`
}

export function compareUpdateTime(a: MetadataInfo, b: IsoDateString): number {
    return new Date(a.updatedAt).getTime() - Date.parse(b)
}
