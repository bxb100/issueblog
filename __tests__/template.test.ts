import {template} from '../src/util/template'

const testData = '{\n' +
    '  "atomLink": "https://github.com/bxb100/issueblog-test/feed.xml",\n' +
    '  "description": "RSS feed of bxb100\'s issueblog-test",\n' +
    '  "link": "https://github.com/bxb100/issueblog-test",\n' +
    '  "title": "bxb100\'s Blog",\n' +
    '  "lastBuildDate": "Wed, 08 Dec 2021 13:46:26 GMT",\n' +
    '  "itunes_image": "https://cdn.jsdelivr.net/gh/bxb100/issueblog-test/blog.png",\n' +
    '  "image": "https://cdn.jsdelivr.net/gh/bxb100/issueblog-test/blog.png",\n' +
    '  "items": [\n' +
    '    {\n' +
    '      "title": "测试文章 2",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sun, 05 Dec 2021 20:09:30 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/22",\n' +
    '      "author": "bxb100",\n' +
    '      "category": "CSSCSCS"\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试一个文章",\n' +
    '      "description": "<h2>\\n<a id=\\"user-content-啦啦啦啦啦\\" class=\\"anchor\\" href=\\"#%E5%95%A6%E5%95%A6%E5%95%A6%E5%95%A6%E5%95%A6\\" aria-hidden=\\"true\\"><span aria-hidden=\\"true\\" class=\\"octicon octicon-link\\"></span></a>啦啦啦啦啦</h2>\\n<p>打的就爱上了</p>\\n<ul>\\n<li>1</li>\\n<li>2</li>\\n<li>3</li>\\n<li>4</li>\\n</ul>\\n<hr>\\n<table>\\n<thead>\\n<tr>\\n<th>a</th>\\n<th>b</th>\\n</tr>\\n</thead>\\n<tbody>\\n<tr>\\n<td>1</td>\\n<td>2</td>\\n</tr>\\n</tbody>\\n</table>\\n",\n' +
    '      "pubDate": "Sun, 05 Dec 2021 20:04:10 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/21",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "测试"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "备份测试 1-改名测试 1",\n' +
    '      "description": "<p>cscscc sc sc sa C</p>\\n",\n' +
    '      "pubDate": "Sun, 05 Dec 2021 12:33:27 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/20",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "todo 测试 2",\n' +
    '      "description": "<ul>\\n<li>[x] 1</li>\\n<li>[x] 2</li>\\n</ul>\\n",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:49:43 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/19",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "Todo"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "todo 测试",\n' +
    '      "description": "<ul>\\n<li>[ ] undone</li>\\n<li>[x] done</li>\\n<li>[x] done2</li>\\n</ul>\\n",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:47:31 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/18",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "Todo"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 7",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:11:30 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/17",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 6",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:11:23 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/16",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 5",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:09:22 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/15",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 4",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:09:13 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/14",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 3",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:06:41 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/13",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 2",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:06:35 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/12",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 1",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:06:29 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/11",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 6",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:01:19 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/10",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 5",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:01:05 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/9",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 4",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:00:54 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/8",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 3",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:00:42 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/7",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 2",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 17:00:31 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/6",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "测试隐藏 label 1",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Sun, 05 Dec 2021 09:45:03 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/5",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "2021",\n' +
    '        "two"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "无 label 测试",\n' +
    '      "description": "<p>cscscs</p>\\n",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 16:05:44 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/4",\n' +
    '      "author": "bxb100",\n' +
    '      "category": []\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "置顶测试 2",\n' +
    '      "description": "<p>测试测试测试测试从 s</p>\\n",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 15:19:37 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/3",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "Top"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "置顶测试 1",\n' +
    '      "description": "<p>cscscs</p>\\n",\n' +
    '      "pubDate": "Sat, 04 Dec 2021 15:19:12 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/2",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "Top"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "友链测试",\n' +
    '      "description": "",\n' +
    '      "pubDate": "Wed, 08 Dec 2021 08:52:45 GMT",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/issues/1",\n' +
    '      "author": "bxb100",\n' +
    '      "category": [\n' +
    '        "Friends"\n' +
    '      ]\n' +
    '    },\n' +
    '    {\n' +
    '      "title": "Ep 34. 和 lepture 聊聊他的产品，以及做个人开发的体验",\n' +
    '      "description": "<p>如果喜欢我们的节目，欢迎通过爱发电打赏支持：<a href=\\"https://afdian.net/@pythonhunter\\" rel=\\"nofollow\\">https://afdian.net/@pythonhunter</a>\\n本期我们和 lepture 聊了聊他的两个项目：Typlog 和 Authlib。lepture 分享了他作为个人开发者的一些经验。\\n本期于 2020 年录制，有些信息可能已经过时。关于这两个项目的情况请以官网为准。</p>\\n<p>嘉宾 lepture\\n主播 Adam Wen laike9m</p>\\n<p>链接</p>\\n<ul>\\n<li>Typlog Authlib Typlog 上的第一个播客 <a href=\\"http://nirokita.cn/\\" rel=\\"nofollow\\">http://nirokita.cn/</a>\\n</li>\\n<li>lepture｜Typlog 作者：中文世界最大的悲剧便是封闭</li>\\n</ul>\\n",\n' +
    '      "link": "https://github.com/bxb100/issueblog-test/releases/tag/ep34",\n' +
    '      "author": "bxb100",\n' +
    '      "pubDate": "Sun, 05 Dec 2021 14:41:32 GMT",\n' +
    '      "enclosure": {\n' +
    '        "url": "https://github.com/bxb100/issueblog-test/releases/download/ep34/8361664978_836589.mp3",\n' +
    '        "length": "39938400",\n' +
    '        "type": "audio/mpeg"\n' +
    '      },\n' +
    '      "category": "Podcast",\n' +
    '      "itunes_item_image": "https://user-images.githubusercontent.com/20685961/144750682-8b5d34c6-f7e1-411d-9d8e-fd30d949cb1b.png"\n' +
    '    }\n' +
    '  ],\n' +
    '  "itunes_author": "bxb100"\n' +
    '}\n'

test('testTemplate', async () => {
  // simple test

  const result = template(JSON.parse(testData))
  console.log(result)
  expect(result).toContain('<![CDATA[测试文章 2]]>')
  expect(result).toContain('<category>测试</category>')
  expect(result).toContain('<guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/20</guid>')
  expect(result).toContain('url="https://github.com/bxb100/issueblog-test/releases/download/ep34/8361664978_836589.mp3"')
  expect(result).toContain('tunes:image href="https://user-images.githubusercontent.com/20685961/144750682-8b5d34c6-f7e1-411d-9d8e-fd30d949cb1b.png"')
  expect(result).toContain('<![CDATA[<p>如果喜欢我们的节目，欢迎通过爱发电打赏支持：')
  expect(result).toContain('<pubDate>Sun, 05 Dec 2021 14:41:32 GMT</pubDate>')
  expect(result).toContain('<category>Podcast</category>')
  expect(result).toContain('<category>two</category>')
  expect(result).toContain('<![CDATA[测试隐藏 label 1]]>')
});
