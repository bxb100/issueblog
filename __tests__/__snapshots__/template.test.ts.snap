// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTemplate 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet href="rss.xsl" type="text/xsl"?>
<rss xmlns:atom="http://www.w3.org/2005/Atom" version="2.0"
     xmlns:itunes="http://www.itunes.com/dtds/podcast-1.0.dtd">
    <channel>
            <title><![CDATA[bxb100's Blog]]></title>
            <link>https://github.com/bxb100/issueblog-test</link>
        <atom:link href="https://github.com/bxb100/issueblog-test/feed.xml" rel="self" type="application/rss+xml" />
            <description><![CDATA[RSS feed of bxb100's issueblog-test]]></description>
        <generator>IssueBlogAction</generator>
        <webMaster>bxb100@outlook.com (@bxb100)</webMaster>
            <itunes:author>bxb100</itunes:author>
                <itunes:explicit>no</itunes:explicit>
                <itunes:image href="https://cdn.jsdelivr.net/gh/bxb100/issueblog-test/blog.png" />
            <language>zh-cn</language>
            <image>
                <url>https://cdn.jsdelivr.net/gh/bxb100/issueblog-test/blog.png</url>
                    <title><![CDATA[bxb100's Blog]]></title>
                <link>https://github.com/bxb100/issueblog-test</link>
            </image>
        <lastBuildDate>Wed, 08 Dec 2021 13:46:26 GMT</lastBuildDate>
            <item>
                <title><![CDATA[测试文章 2]]></title>
                    <description><![CDATA[测试文章 2]]></description>
                    <pubDate>Sun, 05 Dec 2021 20:09:30 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/22</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/22</link>
                    <author><![CDATA[bxb100]]></author>
                    <category>CSSCSCS</category>
            </item>
            <item>
                <title><![CDATA[测试一个文章]]></title>
                    <description><![CDATA[<h2>
<a id="user-content-啦啦啦啦啦" class="anchor" href="#%E5%95%A6%E5%95%A6%E5%95%A6%E5%95%A6%E5%95%A6" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>啦啦啦啦啦</h2>
<p>打的就爱上了</p>
<ul>
<li>1</li>
<li>2</li>
<li>3</li>
<li>4</li>
</ul>
<hr>
<table>
<thead>
<tr>
<th>a</th>
<th>b</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>2</td>
</tr>
</tbody>
</table>
]]></description>
                    <pubDate>Sun, 05 Dec 2021 20:04:10 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/21</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/21</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>测试</category>
            </item>
            <item>
                <title><![CDATA[备份测试 1-改名测试 1]]></title>
                    <description><![CDATA[<p>cscscc sc sc sa C</p>
]]></description>
                    <pubDate>Sun, 05 Dec 2021 12:33:27 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/20</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/20</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[todo 测试 2]]></title>
                    <description><![CDATA[<ul>
<li>[x] 1</li>
<li>[x] 2</li>
</ul>
]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:49:43 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/19</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/19</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>Todo</category>
            </item>
            <item>
                <title><![CDATA[todo 测试]]></title>
                    <description><![CDATA[<ul>
<li>[ ] undone</li>
<li>[x] done</li>
<li>[x] done2</li>
</ul>
]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:47:31 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/18</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/18</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>Todo</category>
            </item>
            <item>
                <title><![CDATA[无 label 7]]></title>
                    <description><![CDATA[无 label 7]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:11:30 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/17</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/17</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 6]]></title>
                    <description><![CDATA[无 label 6]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:11:23 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/16</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/16</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 5]]></title>
                    <description><![CDATA[无 label 5]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:09:22 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/15</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/15</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 4]]></title>
                    <description><![CDATA[无 label 4]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:09:13 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/14</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/14</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 3]]></title>
                    <description><![CDATA[无 label 3]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:06:41 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/13</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/13</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 2]]></title>
                    <description><![CDATA[无 label 2]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:06:35 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/12</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/12</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[无 label 1]]></title>
                    <description><![CDATA[无 label 1]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:06:29 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/11</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/11</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 6]]></title>
                    <description><![CDATA[测试隐藏 label 6]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:01:19 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/10</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/10</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 5]]></title>
                    <description><![CDATA[测试隐藏 label 5]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:01:05 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/9</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/9</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 4]]></title>
                    <description><![CDATA[测试隐藏 label 4]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:00:54 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/8</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/8</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 3]]></title>
                    <description><![CDATA[测试隐藏 label 3]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:00:42 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/7</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/7</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 2]]></title>
                    <description><![CDATA[测试隐藏 label 2]]></description>
                    <pubDate>Sat, 04 Dec 2021 17:00:31 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/6</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/6</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
            </item>
            <item>
                <title><![CDATA[测试隐藏 label 1]]></title>
                    <description><![CDATA[测试隐藏 label 1]]></description>
                    <pubDate>Sun, 05 Dec 2021 09:45:03 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/5</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/5</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>2021</category>
                        <category>two</category>
            </item>
            <item>
                <title><![CDATA[无 label 测试]]></title>
                    <description><![CDATA[<p>cscscs</p>
]]></description>
                    <pubDate>Sat, 04 Dec 2021 16:05:44 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/4</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/4</link>
                    <author><![CDATA[bxb100]]></author>
            </item>
            <item>
                <title><![CDATA[置顶测试 2]]></title>
                    <description><![CDATA[<p>测试测试测试测试从 s</p>
]]></description>
                    <pubDate>Sat, 04 Dec 2021 15:19:37 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/3</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/3</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>Top</category>
            </item>
            <item>
                <title><![CDATA[置顶测试 1]]></title>
                    <description><![CDATA[<p>cscscs</p>
]]></description>
                    <pubDate>Sat, 04 Dec 2021 15:19:12 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/2</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/2</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>Top</category>
            </item>
            <item>
                <title><![CDATA[友链测试]]></title>
                    <description><![CDATA[友链测试]]></description>
                    <pubDate>Wed, 08 Dec 2021 08:52:45 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/issues/1</guid>
                <link>https://github.com/bxb100/issueblog-test/issues/1</link>
                    <author><![CDATA[bxb100]]></author>
                        <category>Friends</category>
            </item>
            <item>
                <title><![CDATA[Ep 34. 和 lepture 聊聊他的产品，以及做个人开发的体验]]></title>
                    <description><![CDATA[<p>如果喜欢我们的节目，欢迎通过爱发电打赏支持：<a href="https://afdian.net/@pythonhunter" rel="nofollow">https://afdian.net/@pythonhunter</a>
本期我们和 lepture 聊了聊他的两个项目：Typlog 和 Authlib。lepture 分享了他作为个人开发者的一些经验。
本期于 2020 年录制，有些信息可能已经过时。关于这两个项目的情况请以官网为准。</p>
<p>嘉宾 lepture
主播 Adam Wen laike9m</p>
<p>链接</p>
<ul>
<li>Typlog Authlib Typlog 上的第一个播客 <a href="http://nirokita.cn/" rel="nofollow">http://nirokita.cn/</a>
</li>
<li>lepture｜Typlog 作者：中文世界最大的悲剧便是封闭</li>
</ul>
]]></description>
                    <pubDate>Sun, 05 Dec 2021 14:41:32 GMT</pubDate>
                        <guid isPermaLink="false">https://github.com/bxb100/issueblog-test/releases/tag/ep34</guid>
                <link>https://github.com/bxb100/issueblog-test/releases/tag/ep34</link>
                        <enclosure url="https://github.com/bxb100/issueblog-test/releases/download/ep34/8361664978_836589.mp3" length="39938400" type="audio/mpeg" />
                        <itunes:image href="https://user-images.githubusercontent.com/20685961/144750682-8b5d34c6-f7e1-411d-9d8e-fd30d949cb1b.png" />
                    <author><![CDATA[bxb100]]></author>
                    <category>Podcast</category>
            </item>
    </channel>
</rss>
"
`;
